vm/vm.go

```diff
@@ -213,32 +213,83 @@ type UnlockedIdentities map[string]*UnlockedIdentity
 
 // SigUnlock performs a signature unlock check and adds the given ident to the set of unlocked identities if
 // the signature is valid, otherwise returns an error.
-func (unlockedIdents UnlockedIdentities) SigUnlock(ident iotago.DirectUnlockableAddress, essence []byte, sig iotago.Signature, inputIndex uint16) error {
+func (unlockedIdents UnlockedIdentities) SigUnlock(ident iotago.DirectUnlockableAddress, essence []byte, sig iotago.Signature, inputIndex uint16, checkUnlockOnly bool) error {
 	if err := ident.Unlock(essence, sig); err != nil {
 		return ierrors.Wrapf(err, "input %d's address is not unlocked through its signature unlock", inputIndex)
 	}
 
+	if checkUnlockOnly {
+		return nil
+	}
+
 	unlockedIdents[ident.Key()] = &UnlockedIdentity{
-		Ident:      ident,
-		UnlockedAt: inputIndex, ReferencedBy: map[uint16]struct{}{},
+		Ident:        ident,
+		UnlockedAt:   inputIndex,
+		ReferencedBy: map[uint16]struct{}{},
 	}
 
 	return nil
 }
 
 // RefUnlock performs a check whether the given ident is unlocked at ref and if so,
 // adds the index of the input to the set of unlocked inputs by this identity.
-func (unlockedIdents UnlockedIdentities) RefUnlock(identKey string, ref uint16, inputIndex uint16) error {
+func (unlockedIdents UnlockedIdentities) RefUnlock(identKey string, ref uint16, inputIndex uint16, checkUnlockOnly bool) error {
 	ident, has := unlockedIdents[identKey]
 	if !has || ident.UnlockedAt != ref {
 		return ierrors.Errorf("input %d is not unlocked through input %d's unlock", inputIndex, ref)
 	}
 
+	if checkUnlockOnly {
+		return nil
+	}
+
 	ident.ReferencedBy[inputIndex] = struct{}{}
 
 	return nil
 }
 
+// MultiUnlock performs a check whether all given unlocks are valid and if so,
+// adds the index of the input to the set of unlocked inputs by this identity.
+func (unlockedIdents UnlockedIdentities) MultiUnlock(vmParams *Params, ident *iotago.MultiAddress, multiUnlock *iotago.MultiUnlock, inputIndex uint16) error {
+	if len(ident.Addresses) != len(multiUnlock.Unlocks) {
+		return ierrors.Wrapf(iotago.ErrMultiAddressAndUnlockLengthDoesNotMatch, "input %d has a multi address (%T) but the amount of addresses does not match the unlocks %d != %d", inputIndex, ident, len(ident.Addresses), len(multiUnlock.Unlocks))
+	}
+
+	var cumulativeUnlockedWeight uint16
+	for subIndex, unlock := range multiUnlock.Unlocks {
+		switch unlock.(type) {
+		case *iotago.EmptyUnlock:
+			// empty unlocks are simply skipped. they are used to maintain correct relationship between
+			// addresses and signatures if the signer doesn't know the particular signature.
```

Comment:
```suggestion
			// Empty unlocks are simply skipped. They are used to maintain correct index relationship between
			// addresses and signatures if the signer doesn't know the signature of another signer.
```

Maybe also add this info to the EmptyUnlock struct where it's more visible?